src/clima_diagnostics.jl

```diff
@@ -80,33 +101,37 @@ function DiagnosticsHandler(scheduled_diagnostics, Y, p, t; dt = nothing)
                 )
             end
         end
+        push!(scheduled_diagnostics_keys, i)
 
         variable = diag.variable
         isa_time_reduction = !isnothing(diag.reduction_time_func)
 
         # The first time we call compute! we use its return value. All the subsequent times
         # (in the callbacks), we will write the result in place
-        storage[diag] = variable.compute!(nothing, Y, p, t)
-        counters[diag] = 1
+        push!(storage, variable.compute!(nothing, Y, p, t))
+        push!(counters, 1)
 
         # If it is not a reduction, call the output writer as well
         if !isa_time_reduction
-            interpolate_field!(diag.output_writer, storage[diag], diag, Y, p, t)
-            write_field!(diag.output_writer, storage[diag], diag, Y, p, t)
+            interpolate_field!(diag.output_writer, storage[i], diag, Y, p, t)
+            write_field!(diag.output_writer, storage[i], diag, Y, p, t)
         else
             # Add to the accumulator
 
             # We use similar + .= instead of copy because CUDA 5.2 does not supported nested
             # wrappers with view(reshape(view)) objects. See discussion in
             # https://github.com/CliMA/ClimaAtmos.jl/pull/2579 and
             # https://github.com/JuliaGPU/Adapt.jl/issues/21
-            accumulators[diag] = similar(storage[diag])
-            accumulators[diag] .= storage[diag]
+            push!(accumulators, similar(storage[i]))
+            accumulators[end] .= storage[i]
         end
     end
+    storage = value_types(storage)[storage...]
+    accumulators = value_types(accumulators)[accumulators...]
```

Comment:
Perhaps you're right, this logic is a bit complicated with the variable length arrays.

My line of thinking was that indexing into `storage` with `storage[diag]`/`storage[i]` should be unchanged, since `storage` is unconditionally extended at every iteration in the `for (i, diag) in enumerate(unique_scheduled_diagnostics)` loop.

The left-hand-side is more tricky, since `accumulators` is not always extended, so I used `push!(accumulators, similar(storage[i]))` for first assigning it, which I think it is correct, because it should refer to the same ith-`storage` object. Then, to be safe, I used `accumulators[end] .= storage[i]`, because `length(accumulators)` may not equal `i`.

Reply:
Actually, we could alternatively keep them as `Dict`s, and just use `Dict{Int, T}`, I think `Dict{Int, Any}` has specialized indexing `Dict{Any}`.

Reply:
(that way, we can let Julia do the book keeping)

Reply:
Or, we could just revert `accumulators` to be a `Dict{Int,T}` and leave `storage` as a vector.

Reply:
Thoughts? preferences? Either way is fine to me, but the `Dict` route does seem a bit easier to me.

Reply:
> Perhaps you're right, this logic is a bit complicated with the variable length arrays.

> My line of thinking was that indexing into storage with storage[diag]/storage[i] should be unchanged, since storage is unconditionally extended at every iteration in the for (i, diag) in enumerate(unique_scheduled_diagnostics) loop.

> The left-hand-side is more tricky, since accumulators is not always extended, so I used push!(accumulators, similar(storage[i])) for first assigning it, which I think it is correct, because it should refer to the same ith-storage object. Then, to be safe, I used accumulators[end] .= storage[i], because length(accumulators) may not equal i.

My concern is mostly when we are indexing in the vectors in the `orchestrate_callback` function. We need to make sure that we are always fetching the correct objects and I am concerned we are not doing that right for `accumulators` because of the variable length. My suggestion is to just always add the accumulators, so just move the 
```
            push!(accumulators, similar(storage[i]))
            accumulators[end] .= storage[i]
```
outside of the `if` statement.

Reply:
> Thoughts? preferences? Either way is fine to me, but the `Dict` route does seem a bit easier to me.

If you think it has no inference/latency implications, go ahead with it. I was going for vectors because of the simpler type structure, but I am happy to keep dicts.

Reply:
> My suggestion is to just always add the accumulators

Wouldn't this mean that we would accumulate more often, though? Or is it that we would just allocate more arrays for accumulation?

Reply:
Ah, I think it'd just be allocating more accumulate arrays, I'll just go with the dict.

Reply:
Yes, it would just allocate more arrays but not use them. Sounds good with going with a dict